File: ios/CropOverlayView.swift
import UIKit

@available(iOS 13.0, *)
class CropOverlayView: UIView {

    var cropRect: CGRect = .zero {
        didSet {
            setNeedsDisplay()
        }
    }

    var allowedRect: CGRect = .zero {
        didSet {
            if cropRect.isEmpty {
                cropRect = allowedRect
            } else {
                let clamped = cropRect.intersection(allowedRect)
                if clamped.isEmpty || clamped.width < minCropSize || clamped.height < minCropSize {
                    cropRect = allowedRect
                }
            }
        }
    }

    var onCropChanged: (() -> Void)?
    var onCropBegan: (() -> Void)?
    var onCropEnded: (() -> Void)?

    var isLightTheme = false {
        didSet { setNeedsDisplay() }
    }

    private let minCropSize: CGFloat = 60
    private let borderWidth: CGFloat = 1.0
    private let cornerLength: CGFloat = 20
    private let cornerWidth: CGFloat = 4.0
    private let edgeHandleLength: CGFloat = 20
    private let gridLineWidth: CGFloat = CGFloat(1.0 / UIScreen.main.scale)
    private let edgeHitZone: CGFloat = 30

    private var activeEdge: DragEdge?
    private var dragStart: CGPoint = .zero
    private var dragStartRect: CGRect = .zero

    private enum DragEdge {
        case top, bottom, left, right
        case topLeft, topRight, bottomLeft, bottomRight
        case move
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }

    private func commonInit() {
        backgroundColor = .clear
        isUserInteractionEnabled = true
        clipsToBounds = false
        isOpaque = false

        let pan = UIPanGestureRecognizer(target: self, action: #selector(handlePan(_:)))
        addGestureRecognizer(pan)

        let pinch = UIPinchGestureRecognizer(target: self, action: #selector(handlePinch(_:)))
        addGestureRecognizer(pinch)
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        guard !cropRect.isEmpty, let ctx = UIGraphicsGetCurrentContext() else { return }
        let cr = cropRect

        let strokeColor = (isLightTheme ? UIColor.black : UIColor.white).cgColor
        ctx.setStrokeColor(strokeColor)
        ctx.setLineWidth(borderWidth)
        ctx.stroke(cr)

        ctx.setLineWidth(gridLineWidth)
        for i in 1...2 {
            let x = cr.minX + cr.width * CGFloat(i) / 3
            ctx.move(to: CGPoint(x: x, y: cr.minY))
            ctx.addLine(to: CGPoint(x: x, y: cr.maxY))
        }
        for i in 1...2 {
            let y = cr.minY + cr.height * CGFloat(i) / 3
            ctx.move(to: CGPoint(x: cr.minX, y: y))
            ctx.addLine(to: CGPoint(x: cr.maxX, y: y))
        }
        ctx.strokePath()

        ctx.setStrokeColor(strokeColor)
        ctx.setLineWidth(cornerWidth)
        ctx.setLineCap(.round)
        ctx.setLineJoin(.round)

        let cl = cornerLength
        let hw = cornerWidth / 2
        let corners: [(CGPoint, CGPoint, CGPoint)] = [
            (CGPoint(x: cr.minX - hw, y: cr.minY + cl),
             CGPoint(x: cr.minX - hw, y: cr.minY - hw),
             CGPoint(x: cr.minX + cl, y: cr.minY - hw)),
            (CGPoint(x: cr.maxX - cl, y: cr.minY - hw),
             CGPoint(x: cr.maxX + hw, y: cr.minY - hw),
             CGPoint(x: cr.maxX + hw, y: cr.minY + cl)),
            (CGPoint(x: cr.minX - hw, y: cr.maxY - cl),
             CGPoint(x: cr.minX - hw, y: cr.maxY + hw),
             CGPoint(x: cr.minX + cl, y: cr.maxY + hw)),
            (CGPoint(x: cr.maxX - cl, y: cr.maxY + hw),
             CGPoint(x: cr.maxX + hw, y: cr.maxY + hw),
             CGPoint(x: cr.maxX + hw, y: cr.maxY - cl)),
        ]
        for (start, corner, end) in corners {
            ctx.move(to: start)
            ctx.addLine(to: corner)
            ctx.addLine(to: end)
        }

        let ehl = edgeHandleLength / 2
        let cx = cr.midX, cy = cr.midY
        ctx.move(to: CGPoint(x: cx - ehl, y: cr.minY - hw))
        ctx.addLine(to: CGPoint(x: cx + ehl, y: cr.minY - hw))
        ctx.move(to: CGPoint(x: cx - ehl, y: cr.maxY + hw))
        ctx.addLine(to: CGPoint(x: cx + ehl, y: cr.maxY + hw))
        ctx.move(to: CGPoint(x: cr.minX - hw, y: cy - ehl))
        ctx.addLine(to: CGPoint(x: cr.minX - hw, y: cy + ehl))
        ctx.move(to: CGPoint(x: cr.maxX + hw, y: cy - ehl))
        ctx.addLine(to: CGPoint(x: cr.maxX + hw, y: cy + ehl))

        ctx.strokePath()
    }

    // MARK: - Hit Testing

    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if detectEdge(at: point) != nil { return self }
        return nil
    }

    private func detectEdge(at pt: CGPoint) -> DragEdge? {
        let r = cropRect
        let z = edgeHitZone

        let nearT = abs(pt.y - r.minY) < z
        let nearB = abs(pt.y - r.maxY) < z
        let nearL = abs(pt.x - r.minX) < z
        let nearR = abs(pt.x - r.maxX) < z
        let inH = pt.x > r.minX - z && pt.x < r.maxX + z
        let inV = pt.y > r.minY - z && pt.y < r.maxY + z

        if nearT && nearL { return .topLeft }
        if nearT && nearR { return .topRight }
        if nearB && nearL { return .bottomLeft }
        if nearB && nearR { return .bottomRight }
        if nearT && inH { return .top }
        if nearB && inH { return .bottom }
        if nearL && inV { return .left }
        if nearR && inV { return .right }
        if r.contains(pt) { return .move }
        return nil
    }

    // MARK: - Gestures

    @objc private func handlePan(_ g: UIPanGestureRecognizer) {
        let pt = g.location(in: self)
        switch g.state {
        case .began:
            activeEdge = detectEdge(at: pt)
            dragStart = pt
            dragStartRect = cropRect
            onCropBegan?()
        case .changed:
            guard let edge = activeEdge else { return }
            let dx = pt.x - dragStart.x
            let dy = pt.y - dragStart.y
            cropRect = computeNewRect(edge: edge, dx: dx, dy: dy)
            onCropChanged?()
        case .ended, .cancelled:
            activeEdge = nil
            onCropEnded?()
        default:
            break
        }
    }

    @objc private func handlePinch(_ g: UIPinchGestureRecognizer) {
        switch g.state {
        case .began:
            dragStartRect = cropRect
            onCropBegan?()
        case .changed:
            let scale = g.scale
            let cx = dragStartRect.midX
            let cy = dragStartRect.midY
            var newW = max(dragStartRect.width * scale, minCropSize)
            var newH = max(dragStartRect.height * scale, minCropSize)
            newW = min(newW, allowedRect.width)
            newH = min(newH, allowedRect.height)
            var r = CGRect(x: cx - newW / 2, y: cy - newH / 2, width: newW, height: newH)
            r = clamp(r, isMove: true)
            cropRect = r
            onCropChanged?()
        case .ended, .cancelled:
            onCropEnded?()
        default:
            break
        }
    }

    // MARK: - Rect Computation

    private func computeNewRect(edge: DragEdge, dx: CGFloat, dy: CGFloat) -> CGRect {
        var r = dragStartRect

        switch edge {
        case .topLeft:
            r.origin.x += dx; r.size.width -= dx
            r.origin.y += dy; r.size.height -= dy
        case .topRight:
            r.size.width += dx
            r.origin.y += dy; r.size.height -= dy
        case .bottomLeft:
            r.origin.x += dx; r.size.width -= dx
            r.size.height += dy
        case .bottomRight:
            r.size.width += dx; r.size.height += dy
        case .top:
            r.origin.y += dy; r.size.height -= dy
        case .bottom:
            r.size.height += dy
        case .left:
            r.origin.x += dx; r.size.width -= dx
        case .right:
            r.size.width += dx
        case .move:
            r.origin.x += dx; r.origin.y += dy
            return clamp(r, isMove: true)
        }

        if r.width < minCropSize {
            let anchorsRight = (edge == .left || edge == .topLeft || edge == .bottomLeft)
            if anchorsRight { r.origin.x = dragStartRect.maxX - minCropSize }
            r.size.width = minCropSize
        }
        if r.height < minCropSize {
            let anchorsBottom = (edge == .top || edge == .topLeft || edge == .topRight)
            if anchorsBottom { r.origin.y = dragStartRect.maxY - minCropSize }
            r.size.height = minCropSize
        }

        return clamp(r, isMove: false)
    }

    private func clamp(_ rect: CGRect, isMove: Bool) -> CGRect {
        var r = rect
        let a = allowedRect
        guard !a.isEmpty else { return r }

        if isMove {
            r.size.width = min(r.width, a.width)
            r.size.height = min(r.height, a.height)
            r.origin.x = max(a.minX, min(r.origin.x, a.maxX - r.width))
            r.origin.y = max(a.minY, min(r.origin.y, a.maxY - r.height))
        } else {
            r.origin.x = max(r.origin.x, a.minX)
            r.origin.y = max(r.origin.y, a.minY)
            if r.maxX > a.maxX { r.size.width = a.maxX - r.origin.x }
            if r.maxY > a.maxY { r.size.height = a.maxY - r.origin.y }
        }
        return r
    }

    func resetCrop() {
        cropRect = allowedRect
    }
}

File: android/src/main/java/com/videotrim/widgets/CropOverlayView.kt
package com.videotrim.widgets

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.View
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class CropOverlayView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

  var cropRect = RectF()
    set(value) {
      field.set(value)
      invalidate()
    }

  var allowedRect = RectF()
    set(value) {
      field.set(value)
      if (cropRect.isEmpty) {
        cropRect = RectF(value)
      } else {
        val clamped = RectF()
        if (!clamped.setIntersect(cropRect, value) || clamped.width() < minCropSize || clamped.height() < minCropSize) {
          cropRect = RectF(value)
        }
      }
    }

  var onCropChanged: (() -> Unit)? = null
  var onCropBegan: (() -> Unit)? = null
  var onCropEnded: (() -> Unit)? = null

  var isLightTheme = false
    set(value) {
      field = value
      val c = if (value) Color.BLACK else Color.WHITE
      borderPaint.color = c
      gridPaint.color = c
      cornerPaint.color = c
      invalidate()
    }

  private val minCropSize = dpToPx(60f)
  private val borderWidth = dpToPx(1f)
  private val cornerLength = dpToPx(20f)
  private val cornerWidth = dpToPx(4f)
  private val edgeHandleLength = dpToPx(20f)
  private val gridLineWidth = 1f / resources.displayMetrics.density
  private val edgeHitZone = dpToPx(30f)

  private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE
    color = Color.WHITE
    strokeWidth = borderWidth
  }
  private val gridPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE
    color = Color.WHITE
    strokeWidth = gridLineWidth
  }
  private val cornerPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE
    color = Color.WHITE
    strokeWidth = cornerWidth
    strokeCap = Paint.Cap.ROUND
    strokeJoin = Paint.Join.ROUND
  }

  private enum class DragEdge {
    TOP, BOTTOM, LEFT, RIGHT,
    TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT,
    MOVE
  }

  private var activeEdge: DragEdge? = null
  private var dragStartX = 0f
  private var dragStartY = 0f
  private var dragStartRect = RectF()
  private var gestureStarted = false

  private val scaleDetector = ScaleGestureDetector(context, object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
    private var pinchStartRect = RectF()

    override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
      pinchStartRect.set(cropRect)
      if (!gestureStarted) {
        gestureStarted = true
        onCropBegan?.invoke()
      }
      return true
    }

    override fun onScale(detector: ScaleGestureDetector): Boolean {
      val scale = detector.scaleFactor
      val cx = pinchStartRect.centerX()
      val cy = pinchStartRect.centerY()
      var newW = max(pinchStartRect.width() * scale, minCropSize)
      var newH = max(pinchStartRect.height() * scale, minCropSize)
      newW = min(newW, allowedRect.width())
      newH = min(newH, allowedRect.height())
      val r = RectF(cx - newW / 2, cy - newH / 2, cx + newW / 2, cy + newH / 2)
      clamp(r, isMove = true)
      cropRect = r
      onCropChanged?.invoke()
      return true
    }

    override fun onScaleEnd(detector: ScaleGestureDetector) {
      if (gestureStarted) {
        gestureStarted = false
        onCropEnded?.invoke()
      }
    }
  })

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    if (cropRect.isEmpty) return
    val cr = cropRect

    canvas.drawRect(cr, borderPaint)

    for (i in 1..2) {
      val x = cr.left + cr.width() * i / 3f
      canvas.drawLine(x, cr.top, x, cr.bottom, gridPaint)
    }
    for (i in 1..2) {
      val y = cr.top + cr.height() * i / 3f
      canvas.drawLine(cr.left, y, cr.right, y, gridPaint)
    }

    val cl = cornerLength
    val hw = cornerWidth / 2f
    val path = Path()
    fun addCorner(sx: Float, sy: Float, cx: Float, cy: Float, ex: Float, ey: Float) {
      path.moveTo(sx, sy); path.lineTo(cx, cy); path.lineTo(ex, ey)
    }
    addCorner(cr.left - hw, cr.top + cl, cr.left - hw, cr.top - hw, cr.left + cl, cr.top - hw)
    addCorner(cr.right - cl, cr.top - hw, cr.right + hw, cr.top - hw, cr.right + hw, cr.top + cl)
    addCorner(cr.left - hw, cr.bottom - cl, cr.left - hw, cr.bottom + hw, cr.left + cl, cr.bottom + hw)
    addCorner(cr.right - cl, cr.bottom + hw, cr.right + hw, cr.bottom + hw, cr.right + hw, cr.bottom - cl)

    val ehl = edgeHandleLength / 2f
    val cx = cr.centerX()
    val cy = cr.centerY()
    path.moveTo(cx - ehl, cr.top - hw); path.lineTo(cx + ehl, cr.top - hw)
    path.moveTo(cx - ehl, cr.bottom + hw); path.lineTo(cx + ehl, cr.bottom + hw)
    path.moveTo(cr.left - hw, cy - ehl); path.lineTo(cr.left - hw, cy + ehl)
    path.moveTo(cr.right + hw, cy - ehl); path.lineTo(cr.right + hw, cy + ehl)

    canvas.drawPath(path, cornerPaint)
  }


  override fun onTouchEvent(event: MotionEvent): Boolean {
    if (event.pointerCount > 1) {
      scaleDetector.onTouchEvent(event)
      return true
    }

    val x = event.x
    val y = event.y

    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        activeEdge = detectEdge(x, y) ?: return false
        dragStartX = x
        dragStartY = y
        dragStartRect.set(cropRect)
        gestureStarted = true
        onCropBegan?.invoke()
        return true
      }
      MotionEvent.ACTION_MOVE -> {
        val edge = activeEdge ?: return false
        val dx = x - dragStartX
        val dy = y - dragStartY
        cropRect = computeNewRect(edge, dx, dy)
        onCropChanged?.invoke()
        return true
      }
      MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
        activeEdge = null
        if (gestureStarted) {
          gestureStarted = false
          onCropEnded?.invoke()
        }
        return true
      }
    }
    return false
  }

  private fun detectEdge(px: Float, py: Float): DragEdge? {
    val r = cropRect
    val z = edgeHitZone

    val nearT = abs(py - r.top) < z
    val nearB = abs(py - r.bottom) < z
    val nearL = abs(px - r.left) < z
    val nearR = abs(px - r.right) < z
    val inH = px > r.left - z && px < r.right + z
    val inV = py > r.top - z && py < r.bottom + z

    if (nearT && nearL) return DragEdge.TOP_LEFT
    if (nearT && nearR) return DragEdge.TOP_RIGHT
    if (nearB && nearL) return DragEdge.BOTTOM_LEFT
    if (nearB && nearR) return DragEdge.BOTTOM_RIGHT
    if (nearT && inH) return DragEdge.TOP
    if (nearB && inH) return DragEdge.BOTTOM
    if (nearL && inV) return DragEdge.LEFT
    if (nearR && inV) return DragEdge.RIGHT
    if (r.contains(px, py)) return DragEdge.MOVE
    return null
  }

  private fun computeNewRect(edge: DragEdge, dx: Float, dy: Float): RectF {
    val r = RectF(dragStartRect)

    when (edge) {
      DragEdge.TOP_LEFT -> { r.left += dx; r.top += dy }
      DragEdge.TOP_RIGHT -> { r.right += dx; r.top += dy }
      DragEdge.BOTTOM_LEFT -> { r.left += dx; r.bottom += dy }
      DragEdge.BOTTOM_RIGHT -> { r.right += dx; r.bottom += dy }
      DragEdge.TOP -> r.top += dy
      DragEdge.BOTTOM -> r.bottom += dy
      DragEdge.LEFT -> r.left += dx
      DragEdge.RIGHT -> r.right += dx
      DragEdge.MOVE -> {
        r.offset(dx, dy)
        clamp(r, isMove = true)
        return r
      }
    }

    if (r.width() < minCropSize) {
      val anchorsRight = edge == DragEdge.LEFT || edge == DragEdge.TOP_LEFT || edge == DragEdge.BOTTOM_LEFT
      if (anchorsRight) r.left = r.right - minCropSize else r.right = r.left + minCropSize
    }
    if (r.height() < minCropSize) {
      val anchorsBottom = edge == DragEdge.TOP || edge == DragEdge.TOP_LEFT || edge == DragEdge.TOP_RIGHT
      if (anchorsBottom) r.top = r.bottom - minCropSize else r.bottom = r.top + minCropSize
    }

    clamp(r, isMove = false)
    return r
  }

  private fun clamp(r: RectF, isMove: Boolean) {
    val a = allowedRect
    if (a.isEmpty) return

    if (isMove) {
      val w = min(r.width(), a.width())
      val h = min(r.height(), a.height())
      r.right = r.left + w
      r.bottom = r.top + h
      if (r.left < a.left) r.offset(a.left - r.left, 0f)
      if (r.top < a.top) r.offset(0f, a.top - r.top)
      if (r.right > a.right) r.offset(a.right - r.right, 0f)
      if (r.bottom > a.bottom) r.offset(0f, a.bottom - r.bottom)
    } else {
      r.left = max(r.left, a.left)
      r.top = max(r.top, a.top)
      if (r.right > a.right) r.right = a.right
      if (r.bottom > a.bottom) r.bottom = a.bottom
    }
  }

  fun resetCrop() {
    cropRect = RectF(allowedRect)
  }

  private fun dpToPx(dp: Float): Float {
    return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, resources.displayMetrics)
  }
}
